libs/Microsoft.MixedReality.WebRTC.Native/src/interop/global_factory.h

```diff
@@ -19,82 +20,112 @@ enum class ObjectType : int {
+  static void SetShutdownOptions(mrsShutdownOptions options) noexcept;
+
+  /// Force-shutdown the library if it is initialized, or does nothing
+  /// otherwise. This call will terminate the WebRTC threads, therefore will
+  /// prevent any dispatched call to a WebRTC object from completing. However,
+  /// by shutting down the threads it will allow unloading the current module
+  /// (DLL), so is recommended to call manually at the end of the process when
+  /// WebRTC objects are not in use anymore but before static deinitializing.
+  /// This is multithread-safe.
   static void ForceShutdown() noexcept;
 
-  GlobalFactory() = default;
-  ~GlobalFactory();
+  /// Attempt to shutdown the library if no tracked object is alive anymore.
+  /// This is always conservative and safe, and will do nothing if any tracked
+  /// object is still alive. The function returns |true| if the library is shut
+  /// down after the call, either because it was already or because the call did
+  /// shut it down. This is multithread-safe.
+  static bool TryShutdown() noexcept;
+
+  /// Try to get a pointer to the (initialized) global factory singleton
+  /// instance. If the library is not initialized, this returns a NULL pointer.
+  /// This is multithread-safe.
+  static RefPtr<GlobalFactory> InstancePtrIfExist() {
+    return GetInstancePtrImpl(/* ensureInitialized = */ false);
+  }
+
+  /// Get a pointer to the global factory singleton instance. If the library is
+  /// not initialized, this call initializes it prior to returning a pointer to
+  /// it. This is multithread-safe.
+  static RefPtr<GlobalFactory> InstancePtr() {
+    return GetInstancePtrImpl(/* ensureInitialized = */ true);
+  }
+
+  /// Add a reference to the library, preventing it from being shutdown. This is
+  /// multithread-safe, and is generally called automatically by |RefPtr<>|.
+  void AddRef() const noexcept {
+    // Calling the member function |AddRef()| implies already holding a
+    // reference, expect for |GetLockImpl()| which will acquire the init mutex
+    // so cannot run concurrently with |RemoveRef()|.
+    RTC_DCHECK(peer_factory_);
+    ref_count_.fetch_add(1, std::memory_order_relaxed);
+  }
+
+  /// Remove a reference to the library acquired with |AddRef()|. If this was
+  /// the last reference, attempt to shutdown the library. This is
+  /// multithread-safe, and is generally called automatically by |RefPtr<>|.
+  void RemoveRef() const noexcept {
+    // Update the reference count under the init lock to ensure it cannot change
+    // from another thread, since invoking |AddRef()| requires having already a
+    // pointer to the GlobalFactory (so this reference would never be the last
+    // one) or calling |GetLockImpl()| to get a new pointer, which will only
+    // call |AddRef()| under the lock too so will block.
+    std::scoped_lock lock(init_mutex_);
+    RTC_DCHECK(peer_factory_);
+    if (ref_count_.fetch_sub(1, std::memory_order_acq_rel) == 1) {

```

Comment:
If this is under the lock does it need to be acq_rel?

Reply:
We still want I think to prevent code from being reordered, especially with the mutex lock above. From a dumb look at it those two lines look independent, so I am not sure the compiler or out-of-order CPU execution would be smart enough to not move the `fetch_sub` outside the mutex lock above.

Reply:
Discussed offline, changed to relaxed, this seems enough.